src/CartesianProduct.jl

```diff
@@ -549,3 +550,36 @@ function vertices_list(cpa::CartesianProductArray{N}
 
     return vlist
 end
+
+
+function linear_map(M::AbstractMatrix{N},
+    cp::CartesianProductArray{N};
+    cond_tol::Number=DEFAULT_COND_TOL,
+    use_inv::Bool=!issparse(M)) where {N<:Real}
+    if !isinvertible(M; cond_tol=cond_tol)
+        error("linear maps for Cartesian product arrays need to be invertible")
+    end
+
+    result = CartesianProductArray(length(cp.array), N)
+    col_st_index, col_end_ind = 0, 0
```

Comment:
I suggest to outsource this preparation step that identifies the starting and ending positions of blocks of a given Cartesian product to its own separate function. It is easier to reason about small functions with a concrete purpose, and this strategy is good for code reuse in general.